ic10emu/src/vm/object/traits.rs

```diff
@@ -0,0 +1,364 @@
+        fn set_source_code_with_invalid(&mut self, code: &str);
+        fn get_source_code(&self) -> String;
+        fn get_line(&self, line: usize) -> Result<&Instruction, ICError>;
+    }
+
+    pub trait CircuitHolder: Logicable + Storage {
+        fn clear_error(&mut self);
+        fn set_error(&mut self, state: i32);
+        /// i32::MAX is db
+        fn get_logicable_from_index(
+            &self,
+            device: i32,
+            connection: Option<usize>,
+        ) -> Option<ObjectRef>;
+        /// i32::MAX is db
+        fn get_logicable_from_index_mut(
+            &mut self,
+            device: i32,
+            connection: Option<usize>,
+        ) -> Option<ObjectRefMut>;
+        fn get_logicable_from_id(
+            &self,
+            device: ObjectID,
+            connection: Option<usize>,
+        ) -> Option<ObjectRef>;
+        fn get_logicable_from_id_mut(
+            &mut self,
+            device: ObjectID,
+            connection: Option<usize>,
+        ) -> Option<ObjectRefMut>;
+        fn get_ic(&self) -> Option<VMObject>;
+        fn hault_and_catch_fire(&mut self);
+    }
+
+    pub trait Item {
+        fn consumable(&self) -> bool;
+        fn filter_type(&self) -> Option<GasType>;
+        fn ingredient(&self) -> bool;
+        fn max_quantity(&self) -> u32;
+        fn reagents(&self) -> Option<&BTreeMap<String, f64>>;
+        fn slot_class(&self) -> Class;
+        fn sorting_class(&self) -> SortingClass;
+        fn get_parent_slot(&self) -> Option<ParentSlotInfo>;
+        fn set_parent_slot(&mut self, info: Option<ParentSlotInfo>);
+        fn get_damage(&self) -> f32;
+        fn set_damage(&mut self, damage: f32);
+        fn root_parent_human(&self) -> Option<VMObject> {
+            self.get_parent_slot().and_then(|info| {
+                if let Some(obj) = self.get_vm().get_object(info.parent) {
+                    if obj.borrow().as_human().is_some() {
+                        return Some(obj);
+                    }
+                    let obj_ref = obj.borrow();
+                    if let Some(item) = obj_ref.as_item() {
+                        return item.root_parent_human()
+                    }
+                }
+                None
+            })
+        }
+    }
+
+    pub trait Plant {
+        fn get_efficiency(&self) -> f64;
+        fn get_health(&self) -> f64;
+        fn get_growth(&self) -> f64;
+        fn is_mature(&self) -> bool;
+        fn is_seeding(&self) -> bool;
+    }
+
+    pub trait Suit: Item + Storage {
+        fn pressure_waste(&self) -> f32;
+        fn pressure_waste_max(&self) -> f32;
+        fn pressure_air(&self) -> f32;
+    }
+
+    pub trait InternalAtmosphere {
+        fn get_volume(&self) -> f64;
+    }
+
+    pub trait Thermal {
+        fn get_convection_factor(&self) -> f32;
+        fn get_radiation_factor(&self) -> f32;
+    }
+
+    pub trait IntegratedCircuit: Logicable + MemoryWritable + SourceCode + Item {
```

Comment:
some documentation on these functions would be nice

Reply:
Is this what you were looking for? or should I be more in depth?

Reply:
looks good to me!